File: 8.string-to-integer-atoi.java
import javax.crypto.spec.ChaCha20ParameterSpec;

/*
 * @lc app=leetcode id=8 lang=java
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
class Solution {
    public int myAtoi(String str) {
        if (str.length() == 0)
            return 0;
        int index = 0;
        int answer = 0;
        int sign = 1;
        char num = str.charAt(index);
        while (!Character.isDigit(num)) {
            if (Character.isWhitespace(num)) {
                index++;
                if (index >= str.length())
                    return answer;
                num = str.charAt(index);
                continue;
            }
            if (Character.isLetter(num) || num == '.')
                return answer;
            if (num == '-') {
                index++;
                sign = -1;
                break;
            } else if (num == '+') {
                index++;
                break;
            }
            if (index >= str.length())
                return answer;
            num = str.charAt(index);
        }
        if (index >= str.length())
            return answer;
        if (!Character.isDigit(str.charAt(index)))
            return 0;
        while (str.charAt(index) == '0') {
            index++;
            if (index >= str.length())
                return answer;
        }

        if (Character.isDigit(str.charAt(index)))
            answer += str.charAt(index) - '0';
        else
            return answer;
        index++;
        for (int count = 0; count < 8; count++) {
            if (index >= str.length()) {
                return answer * sign;
            }
            num = str.charAt(index);
            if (!Character.isDigit(num)) {
                return answer * sign;
            }
            answer = 10 * answer + num - '0';
            index++;
        }
        if (index >= str.length())
            return answer * sign;
        num = str.charAt(index);
        if (!Character.isDigit(num)) {
            return answer * sign;
        }
        boolean overflow = answer > Integer.MAX_VALUE / 10;
        if (overflow) {
            if (sign == 1) {
                return Integer.MAX_VALUE;
            } else {
                return Integer.MIN_VALUE;
            }
        }
        answer = answer * 10 + num - '0';
        if (answer < 0) {
            if (sign == 1) {
                return Integer.MAX_VALUE;
            } else {
                return Integer.MIN_VALUE;
            }
        }
        index++;
        if (index >= str.length())
            return answer * sign;
        num = str.charAt(index);
        if (Character.isDigit(num)) {
            if (sign == 1) {
                return Integer.MAX_VALUE;
            } else {
                return Integer.MIN_VALUE;
            }
        }
        return answer * sign;
    }
}
// @lc code=end

File: csharp/medium/8.string-to-integer-atoi.cs
/*
 * @lc app=leetcode id=8 lang=csharp
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
public class Solution
{
    enum State { Start, Signed, InNumber, End };
    Dictionary<(State, int), State> dic;

    public Solution() {
        dic = new Dictionary<(State, int), State>();
        dic.Add((State.Start, 0), State.Start);
        dic.Add((State.Start, 1), State.Signed);
        dic.Add((State.Start, 2), State.InNumber);
        dic.Add((State.Start, 3), State.End);
        dic.Add((State.Signed, 0), State.End);
        dic.Add((State.Signed, 1), State.End);
        dic.Add((State.Signed, 2), State.InNumber);
        dic.Add((State.Signed, 3), State.End);
        dic.Add((State.InNumber, 0), State.End);
        dic.Add((State.InNumber, 1), State.End);
        dic.Add((State.InNumber, 2), State.InNumber);
        dic.Add((State.InNumber, 3), State.End);
        dic.Add((State.End, 0), State.End);
        dic.Add((State.End, 1), State.End);
        dic.Add((State.End, 2), State.End);
        dic.Add((State.End, 3), State.End);
    }
    public int MyAtoi(string str) {
        long res = 0;
        int sign = 1;
        var state = State.Start;
        foreach(var ch in str) {
            state = dic[(state, GetAlpha(ch))];
            if(state == State.InNumber) {
                res = res * 10 + ch - '0';
                res = sign == 1 ? Math.Min(res, (long)int.MaxValue) : Math.Min(res, -(long)int.MinValue);
            }
            else if(state == State.Signed)
                sign = ch == '+' ? 1 : -1;
        }
        return (int)(sign * res);
    }

    private int GetAlpha(char ch) {
        if(char.IsWhiteSpace(ch)) return 0;
        else if(ch == '+' || ch == '-') return 1;
        else if(char.IsNumber(ch)) return 2;
        else return 3;
    }
}
// @lc code=end

